snapshot-cross-account-replicator/functions/lambda.py

```diff
@@ -113,70 +204,140 @@ def cleanup_snapshots(event, context):
+                    print('Replicating snapshot ' +
+                          snapshot['DBClusterSnapshotIdentifier'] + ' to region ' + target_region)
+                else:
+                    print('Replicating snapshot ' +
+                          snapshot['DBSnapshotIdentifier'] + ' to region ' + target_region)
+                target_region_rds = boto3.client(
+                    'rds', region_name=target_region)
+                copy_snapshot(snapshot, target_region_rds, source_region)
+            elif replication_type == 'cross-account':
+                replicate_snapshot_cross_account(rds, snapshot)
 
 
 def create_snapshots(event, context):
-    """Lambda entry point for the snapshot creation"""
+    """Lambda entry point for the snapshot creation (STEP1)"""
 
     print('Lambda function start: going to create snapshots for the RDS instances ' + instances)
 
     source_rds = boto3.client('rds', region_name=source_region)
 
+    tags = [
+        {
+            'Key': 'created_by',
+            'Value': setup_name
+        },
+        {
+            'Key': 'replicating',
+            'Value': 'false'
+        }
+    ]
+
     for instance in instances.split(','):
         now = datetime.datetime.now()
         db_snapshot_name = instance + '-' + now.strftime('%Y-%m-%d-%H-%M')
         try:
-            source_rds.create_db_snapshot(
-                DBSnapshotIdentifier=db_snapshot_name,
-                DBInstanceIdentifier=instance,
-                Tags=[
-                    {
-                        'Key': 'created_by',
-                        'Value': setup_name
-                    }
-                ])
+            if is_cluster:
+                source_rds.create_db_cluster_snapshot(
+                    DBClusterSnapshotIdentifier=db_snapshot_name,
+                    DBClusterIdentifier=instance,
+                    Tags=tags
+                )
+            else:
+                source_rds.create_db_snapshot(
+                    DBSnapshotIdentifier=db_snapshot_name,
+                    DBInstanceIdentifier=instance,
+                    Tags=tags
+                )
         except botocore.exceptions.ClientError as e:
             raise Exception("Could not issue create command: %s" % e)
 
 
 def delete_intermediate_snapshot(event, context):
-    """Lambda entry point for cleaning up the 2nd intermediate snapshot"""
+    """Lambda entry point for cleaning up the 2nd intermediate snapshot (STEP4)"""
 
     rds = boto3.client('rds', region_name=target_region)
-    snapshot = match_snapshot_event(rds, event)
-    if snapshot:
-        source_rds = get_assumed_role_rds_client(
-            source_account_iam_role_arn, target_region)
-        delete_snapshot(source_rds, snapshot['DBSnapshotIdentifier'])
+    source_rds = get_assumed_role_rds_client(
+        source_account_iam_role_arn, target_region)
+
+    # CRON based, used for step 4 (cluster)
+    if is_cluster:
+        snapshots = match_cluster_snapshots(rds)
```

Comment:
hmmm this will never match the intermediate snapshot right? So when the step 3 function is called, it sets the `replicating` tag to `true` in the intermediate snapshot, and deletes the initial one. But then when this function is called (step 4), the `match_tags` function won't match the intermediate snapshot as it has the `replicating` tag to `true`. Am I right? So the intermediate won't ever be deleted

Reply:
Not true, because it uses the target-account snapshot for matching (which has `replicating=fase`, and if status is available, so the copy operation is done), then deletes the equivalent intermediate snapshot.

Reply:
right. Ok nvm then :)